File: Dynamic_Programming/20_Oct_2019_binomial_coefficient/20_Oct_2019_binomial_coefficient_dynamic.cpp
#include<bits/stdc++.h>

using namespace std;

//return min of2 ints
int min(int a, int b){
    return (a>b)? a:b;
}

int binomialCoeff(int n, int k){
    int c[n+1][k+1];
    int i, j;

    for (i = 0; i <=  n ; i++){
        for(j = 0; j <= min(i,k); j++){
            //base case
            if(j == 0 || j == i){
                c[i][j] = 1;
            }
            else{
                c[i][j] = c[i-1][j-1] + c[i-1][j];
            }
            
        }
    }
    return c[n][k];
}

//driver code
int main(){
    int n = 5, k = 2;

    cout << "Value of C("<<n<<", "<<k<<") is "<< binomialCoeff(n,k);
    return 0;
}

File: Dynamic_Programming/20_Oct_2019_binomial_coefficient/20_Oct_2019_binomial_coefficient_dynamic.py
#dynamic binomial coeff

def binomialCoeff(n,k):
    c = [[0 for i in range(k+1)] for i in range(n+1)]

    for i in range(n+1):
        for j in range(min(i,k)+1):
            if j == 0 or j == i:
                c[i][j] = 1
            else:
                c[i][j] = c[i-1][j-1] + c[i-1][j]
    
    return c[n][k]

#driver code
n = 5
k = 2

print("The binomial coeff of {} and {} is {}".format(n,k,binomialCoeff(n,k)))